File: python/examples/vtk/csg_fracture.py
"""
Cellular fracture: one arrangement of a closed mesh + a grid of axis-aligned
cutting planes, split into its volumetric box cells, coloured and optionally
exploded. Showcases volumetric-domain extraction (not a boolean) from a single
build: tf.CsgGraph builds the arrangement once, and the interior cells are
read directly as the domains inside operand 0 — graph.domains(tf.op(0)).

  python csg_fracture.py MESH [cuts] [explode] [--all] [--png]

  MESH        path to a closed surface mesh (stl / obj)
  cuts        number of axis-aligned cuts per axis              [default 12]
  explode     outward cell displacement, fraction of extent     [default 0.0]
  --all       render every domain, skipping the interior filter
  --png       render offscreen to /tmp/fracture.png

Interactive controls: drag orbit, scroll zoom, 'c' print camera, 'q' quit.
"""
import os
import numpy as np
import vtk
import trueform as tf
from util import numpy_to_polydata, create_parser

DT, IDT = np.float64, np.int32

# Pastel qualitative palette derived from the paper's accents (tfteal/tfrose
# brand + cOne..cFive Tableau set), softened toward the light-teal/light-rose
# domain figures so this plate reads as part of the same family.
PALETTE = [
    "8FE3D6",  # soft teal   (tfteal / aFace)
    "EFA9C4",  # soft rose   (tfrose / bFace)
    "8FB4D8",  # soft blue   (cOne)
    "F6BE84",  # soft orange (cTwo)
    "8FC882",  # soft green  (cThree)
    "B9A0DC",  # soft purple (cFour)
    "EF9A98",  # soft red    (cFive)
    "6FC3B4",  # deep teal   (tfd2face)
    "F2CE86",  # soft gold
    "A9C4E6",  # soft sky
    "A6DDBE",  # soft mint
    "D6A9CE",  # soft mauve
]
EDGE_RGB = (0.18, 0.20, 0.24)   # near-ink cell outlines (tfink family)

# Camera override for reproducing a hand-picked view. Press 'c' in the
# interactive window to print a ready-to-paste block, drop it here, then a
# --png render replicates that exact view. None -> default ears-up framing.
#   CAM = dict(pos=(...), foc=(...), up=(...), angle=..., size=(W, H))
CAM = None


def cell_color(i):
    hx = PALETTE[(i * 5) % len(PALETTE)]   # coprime step spreads neighbours
    return tuple(int(hx[k:k + 2], 16) / 255 for k in (0, 2, 4))


def load_closed(path):
    ext = os.path.splitext(path)[1].lower()
    if ext == ".stl":
        faces, pts = tf.read_stl(path, index_dtype=IDT)
    elif ext == ".obj":
        faces, pts = tf.read_obj(path, ngon=3, dtype=DT, index_dtype=IDT)
    else:
        raise SystemExit(f"unsupported mesh extension '{ext}' (use .stl or .obj)")
    faces, pts = tf.cleaned((faces, pts.astype(DT)), tolerance=1e-6)
    if not tf.is_closed(tf.Mesh(faces, pts)):
        raise SystemExit(f"{os.path.basename(path)} is not watertight; "
                         "'inside the solid' is undefined on an open surface")
    # normalise to unit max-extent, centred at the origin
    c = (pts.min(0) + pts.max(0)) / 2
    ext = (pts.max(0) - pts.min(0)).max()
    return faces, (pts - c) / ext


def rot_z_to(n):
    z = np.array([0, 0, 1.0])
    if np.allclose(n, z): return np.eye(3)
    if np.allclose(n, -z): return np.diag([1.0, -1.0, -1.0])
    ax = np.cross(z, n); ax /= np.linalg.norm(ax)
    co = float(np.dot(z, n)); si = np.sqrt(max(0.0, 1 - co * co))
    K = np.array([[0, -ax[2], ax[1]], [ax[2], 0, -ax[0]], [-ax[1], ax[0], 0]])
    return np.eye(3) + si * K + (1 - co) * (K @ K)


def make_plane(center, normal, side=4.0):
    pf, pp = tf.make_plane_mesh(side, side, dtype=DT, index_dtype=IDT)
    pp = np.asarray(pp, DT)
    pp = (pp - pp.mean(0)) @ rot_z_to(normal).T + center
    return tf.Mesh(np.asarray(pf, IDT), pp)


def make_planes(bp, n):
    # n axis-aligned cuts per axis at evenly spaced interior offsets -> box cells
    lo, hi = bp.min(0), bp.max(0)
    mid = (lo + hi) / 2
    planes = []
    for axis in range(3):
        nrm = np.zeros(3); nrm[axis] = 1.0
        for i in range(n):
            t = (i + 1) / (n + 1)
            ctr = mid.copy(); ctr[axis] = lo[axis] + t * (hi[axis] - lo[axis])
            planes.append(make_plane(ctr, nrm))
    return planes


def set_camera(ren, win, cam):
    if CAM:
        cam.SetPosition(*CAM["pos"]); cam.SetFocalPoint(*CAM["foc"])
        cam.SetViewUp(*CAM["up"]); cam.SetViewAngle(CAM["angle"])
        win.SetSize(*CAM["size"])
    else:
        # explicit direction so z is genuinely up (avoids the parallel view-up
        # reset); look from the front (-y), to the side (+x) and above.
        win.SetSize(1400, 1400)
        cam.SetViewUp(0, 0, 1)
        cam.SetFocalPoint(0, 0, 0)
        cam.SetPosition(0.55, -1.0, 0.32)
        ren.ResetCamera()
        cam.Zoom(1.35)
    ren.ResetCameraClippingRange()


def main():
    parser = create_parser(__doc__, mesh_args=1)
    parser.add_argument("cuts", nargs="?", type=int, default=12,
                        help="number of axis-aligned cuts per axis")
    parser.add_argument("explode", nargs="?", type=float, default=0.0,
                        help="outward cell displacement, fraction of extent")
    parser.add_argument("--all", action="store_true", dest="all_domains",
                        help="render every domain, skip the interior filter")
    parser.add_argument("--png", action="store_true",
                        help="render offscreen to /tmp/fracture.png")
    args = parser.parse_args()

    sf, sp = load_closed(args.mesh)
    name = os.path.basename(args.mesh)
    print(f"{name}: {len(sf)} tris, bbox {sp.min(0)} .. {sp.max(0)}")

    meshes = [tf.Mesh(sf, sp)] + make_planes(sp, args.cuts)

    # One build; every query after this reuses it.
    graph = tf.CsgGraph(meshes)
    if args.all_domains:
        domains, _ids = graph.domains()
    else:
        # a cell is interior iff it lies inside operand 0 (the solid)
        domains, _ids = graph.domains(tf.op(0))
    print(f"cells: {len(domains)}" + ("  (ALL)" if args.all_domains else
                                      "  (interior of solid)"))

    scen = sp.mean(0)
    cells = []
    for cf, cp in domains:
        cp = np.asarray(cp, DT); cf = np.asarray(cf)
        if len(cp) == 0 or len(cf) < 1: continue
        cells.append((cf, cp + args.explode * (cp.mean(0) - scen)))
    print(f"rendered domains: {len(cells)}")

    ren = vtk.vtkRenderer(); ren.SetBackground(1, 1, 1)
    if not args.png:
        ren.UseFXAAOn()   # FXAA halos transparent edges; rely on MSAA for PNG
    for i, (cf, cp) in enumerate(cells):
        poly = numpy_to_polydata(cp.astype(np.float64), cf)
        m = vtk.vtkPolyDataMapper(); m.SetInputData(poly)
        a = vtk.vtkActor(); a.SetMapper(m)
        p = a.GetProperty()
        p.SetColor(*cell_color(i))
        p.SetEdgeVisibility(0)            # no triangulation
        p.SetInterpolationToPhong()
        p.SetAmbient(0.32); p.SetDiffuse(0.68)
        p.SetSpecular(0.18); p.SetSpecularPower(28)
        ren.AddActor(a)
        # crisp feature edges only (cut seams + silhouette), not the mesh
        fe = vtk.vtkFeatureEdges(); fe.SetInputData(poly)
        fe.BoundaryEdgesOn(); fe.FeatureEdgesOn(); fe.SetFeatureAngle(25)
        fe.ManifoldEdgesOff(); fe.NonManifoldEdgesOff(); fe.Update()
        em = vtk.vtkPolyDataMapper(); em.SetInputConnection(fe.GetOutputPort())
        em.ScalarVisibilityOff()
        ea = vtk.vtkActor(); ea.SetMapper(em)
        ea.GetProperty().SetColor(*EDGE_RGB)
        ea.GetProperty().SetLineWidth(1.1)
        ea.GetProperty().LightingOff()
        ren.AddActor(ea)

    # studio 3-point lighting for shape-revealing shading
    lk = vtk.vtkLightKit()
    lk.SetKeyLightIntensity(0.9)
    lk.SetKeyToFillRatio(2.2); lk.SetKeyToHeadRatio(3.5); lk.SetKeyToBackRatio(3.0)
    lk.AddLightsToRenderer(ren)

    win = vtk.vtkRenderWindow()
    win.SetMultiSamples(8)
    win.AddRenderer(ren)
    win.SetWindowName(f"csg_fracture: {name} / {len(cells)} cells / {args.cuts} cuts")
    cam = ren.GetActiveCamera()
    set_camera(ren, win, cam)

    if args.png:
        win.SetAlphaBitPlanes(1)          # transparent background (RGBA)
        win.SetOffScreenRendering(1)
        win.Render()
        w2i = vtk.vtkWindowToImageFilter(); w2i.SetInput(win)
        w2i.SetInputBufferTypeToRGBA(); w2i.ReadFrontBufferOff(); w2i.Update()
        wr = vtk.vtkPNGWriter(); wr.SetFileName("/tmp/fracture.png")
        wr.SetInputConnection(w2i.GetOutputPort()); wr.Write()
        print("wrote /tmp/fracture.png (RGBA)")
    else:
        iren = vtk.vtkRenderWindowInteractor()
        iren.SetRenderWindow(win)
        iren.SetInteractorStyle(vtk.vtkInteractorStyleTrackballCamera())

        def dump_camera(obj, _evt):
            if obj.GetKeySym() not in ("c", "C"):
                return
            rd = lambda v: tuple(round(x, 4) for x in v)
            print("\nCAM = dict(")
            print(f"    pos={rd(cam.GetPosition())},")
            print(f"    foc={rd(cam.GetFocalPoint())},")
            print(f"    up={rd(cam.GetViewUp())},")
            print(f"    angle={round(cam.GetViewAngle(), 3)}, "
                  f"size={tuple(win.GetSize())})\n")
        iren.AddObserver("KeyPressEvent", dump_camera)

        win.Render()
        print("interactive: drag orbit, scroll zoom, 'c' print camera, 'q' quit")
        iren.Start()


if __name__ == "__main__":
    main()

File: cpp/examples/vtk/csg_fracture.cpp
/*
 * Cellular fracture from one reusable CSG arrangement.
 *
 * Usage: csg_fracture MESH [cuts] [explode] [--all] [--png]
 *   cuts    axis-aligned cuts per axis                         [default 12]
 *   explode outward displacement as a fraction of extent       [default 0.0]
 *   --all   render every bounded domain instead of operand 0's interior
 *   --png   render RGBA offscreen to the temporary fracture.png path
 */
#include "util/vtk_examples.hpp"

#include <trueform/cpp/clean.hpp>
#include <trueform/cpp/core/cache.hpp>
#include <trueform/cpp/core/index_type.hpp>
#include <trueform/cpp/core/mesh.hpp>
#include <trueform/cpp/csg/csg_domains.hpp>
#include <trueform/cpp/csg/csg_graph.hpp>
#include <trueform/cpp/geometry/make_plane_mesh.hpp>
#include <trueform/cpp/io/read_obj.hpp>
#include <trueform/cpp/io/read_stl.hpp>
#include <trueform/cpp/topology.hpp>

#include <vtkCamera.h>
#include <vtkErrorCode.h>
#include <vtkFeatureEdges.h>
#include <vtkInteractorStyleTrackballCamera.h>
#include <vtkLightKit.h>
#include <vtkObjectFactory.h>
#include <vtkPNGWriter.h>
#include <vtkPolyDataMapper.h>
#include <vtkProperty.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkRenderer.h>
#include <vtkWindowToImageFilter.h>

#include <algorithm>
#include <array>
#include <cctype>
#include <cmath>
#include <cstdint>
#include <exception>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace {

namespace examples = trueform::vtk_examples;
using index_t = tf::cpp::default_index_t;
using storage = tf::polygons_buffer<index_t, double, 3, 3>;
using cache_t = tf::cpp::cache<index_t, double>;
using mesh_t = tf::cpp::mesh<index_t, double>;
using point = examples::point3<double>;

constexpr std::array<std::array<unsigned char, 3>, 12> palette{{
    {0x8F, 0xE3, 0xD6},
    {0xEF, 0xA9, 0xC4},
    {0x8F, 0xB4, 0xD8},
    {0xF6, 0xBE, 0x84},
    {0x8F, 0xC8, 0x82},
    {0xB9, 0xA0, 0xDC},
    {0xEF, 0x9A, 0x98},
    {0x6F, 0xC3, 0xB4},
    {0xF2, 0xCE, 0x86},
    {0xA9, 0xC4, 0xE6},
    {0xA6, 0xDD, 0xBE},
    {0xD6, 0xA9, 0xCE},
}};
constexpr std::array<double, 3> edge_rgb{0.18, 0.20, 0.24};

struct camera_settings {
  point position;
  point focal_point;
  point view_up;
  double view_angle;
  std::array<int, 2> size;
};

// Replace nullopt with values printed by the C key to reproduce a chosen view.
const std::optional<camera_settings> camera_override = std::nullopt;

struct options {
  std::filesystem::path mesh_path;
  int cuts = 12;
  double explode = 0.0;
  bool all_domains = false;
  bool png = false;
};

auto lower(std::string value) -> std::string {
  std::transform(
      value.begin(), value.end(), value.begin(),
      [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
  return value;
}

auto parse_int(const std::string &text) -> int {
  std::size_t consumed = 0;
  const auto value = std::stoi(text, &consumed);
  if (consumed != text.size())
    throw std::invalid_argument("invalid cuts value '" + text + "'");
  return value;
}

auto parse_double(const std::string &text) -> double {
  std::size_t consumed = 0;
  const auto value = std::stod(text, &consumed);
  if (consumed != text.size())
    throw std::invalid_argument("invalid explode value '" + text + "'");
  return value;
}

void print_usage(const char *program, std::ostream &stream) {
  stream
      << "Usage: " << program << " MESH [cuts] [explode] [--all] [--png]\n\n"
      << "  MESH     closed .stl or .obj surface\n"
      << "  cuts     number of axis-aligned cuts per axis [default 12]\n"
      << "  explode  outward cell displacement fraction  [default 0.0]\n"
      << "  --all    render every domain, skip the interior filter\n"
      << "  --png    render offscreen to the temporary fracture.png path\n\n"
      << "Interactive: drag orbit, scroll zoom, 'c' print camera, 'q' quit.\n";
}

auto parse_options(int argc, char **argv) -> std::optional<options> {
  options result;
  std::vector<std::string> positional;
  for (int index = 1; index != argc; ++index) {
    const auto argument = std::string(argv[index]);
    if (argument == "--all")
      result.all_domains = true;
    else if (argument == "--png")
      result.png = true;
    else if (argument == "--help" || argument == "-h") {
      print_usage(argv[0], std::cout);
      return std::nullopt;
    } else if (argument.rfind("--", 0) == 0)
      throw std::invalid_argument("unknown option '" + argument + "'");
    else
      positional.push_back(argument);
  }
  if (positional.empty())
    throw std::invalid_argument("MESH is required");
  if (positional.size() > 3)
    throw std::invalid_argument("too many positional arguments");
  result.mesh_path = positional[0];
  if (positional.size() > 1)
    result.cuts = parse_int(positional[1]);
  if (positional.size() > 2)
    result.explode = parse_double(positional[2]);
  return result;
}

auto stl_as_double(const std::filesystem::path &path) -> storage {
  const auto source = tf::cpp::read_stl(path.string());
  const auto &source_coordinates = source.points_buffer().data_buffer();
  storage result;
  result.faces_buffer().data_buffer() = source.faces_buffer().data_buffer();
  auto &coordinates = result.points_buffer().data_buffer();
  coordinates.allocate(source_coordinates.size());
  std::copy(source_coordinates.begin(), source_coordinates.end(),
            coordinates.begin());
  return result;
}

auto load_closed(const std::filesystem::path &path) -> storage {
  const auto extension = lower(path.extension().string());
  storage source;
  if (extension == ".stl")
    source = stl_as_double(path);
  else if (extension == ".obj")
    source = tf::cpp::read_obj<index_t, double, 3>(path.string());
  else
    throw std::invalid_argument("unsupported mesh extension '" + extension +
                                "' (use .stl or .obj)");

  cache_t source_cache;
  auto value =
      tf::cpp::cleaned_mesh(examples::mesh_over(source, source_cache), 1.0e-6);
  cache_t cache;
  if (!tf::cpp::is_closed(examples::mesh_over(value, cache)))
    throw std::invalid_argument(path.filename().string() +
                                " is not watertight; 'inside the solid' is "
                                "undefined on an open surface");

  const auto [lower_bound, upper_bound] =
      examples::point_bounds(value.points_buffer());
  const point center{(lower_bound[0] + upper_bound[0]) / 2.0,
                     (lower_bound[1] + upper_bound[1]) / 2.0,
                     (lower_bound[2] + upper_bound[2]) / 2.0};
  const auto extent = std::max({upper_bound[0] - lower_bound[0],
                                upper_bound[1] - lower_bound[1],
                                upper_bound[2] - lower_bound[2]});
  if (!(extent > 0.0))
    throw std::invalid_argument("mesh has zero extent");
  // The coordinates move after the last read of them, so nothing the cache
  // holds is stale by the time this is handed back.
  for (auto vertex : value.points())
    for (std::size_t axis = 0; axis != 3; ++axis)
      vertex[axis] = (vertex[axis] - center[axis]) / extent;
  return value;
}

using matrix3 = std::array<double, 9>;

auto rotation_z_to(const point &normal) -> matrix3 {
  if (std::abs(normal[0]) < 1.0e-12 && std::abs(normal[1]) < 1.0e-12 &&
      std::abs(normal[2] - 1.0) < 1.0e-12)
    return {1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0};
  if (std::abs(normal[0]) < 1.0e-12 && std::abs(normal[1]) < 1.0e-12 &&
      std::abs(normal[2] + 1.0) < 1.0e-12)
    return {1.0, 0.0, 0.0, 0.0, -1.0, 0.0, 0.0, 0.0, -1.0};

  point axis{-normal[1], normal[0], 0.0};
  const auto axis_length = std::sqrt(axis[0] * axis[0] + axis[1] * axis[1]);
  for (auto &coordinate : axis)
    coordinate /= axis_length;
  const auto cosine = normal[2];
  const auto sine = std::sqrt(std::max(0.0, 1.0 - cosine * cosine));
  const matrix3 k{0.0,      -axis[2], axis[1], axis[2], 0.0,
                  -axis[0], -axis[1], axis[0], 0.0};
  matrix3 kk{};
  for (int row = 0; row != 3; ++row)
    for (int column = 0; column != 3; ++column)
      for (int inner = 0; inner != 3; ++inner)
        kk[static_cast<std::size_t>(row * 3 + column)] +=
            k[static_cast<std::size_t>(row * 3 + inner)] *
            k[static_cast<std::size_t>(inner * 3 + column)];
  matrix3 result{1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0};
  for (std::size_t index = 0; index != result.size(); ++index)
    result[index] += sine * k[index] + (1.0 - cosine) * kk[index];
  return result;
}

auto make_plane(const point &center, const point &normal, double side = 4.0)
    -> storage {
  auto value = tf::cpp::make_plane_mesh(side, side);
  const auto rotation = rotation_z_to(normal);
  const auto local_center = examples::point_mean(value.points_buffer());
  for (auto vertex : value.points()) {
    const point local{vertex[0] - local_center[0], vertex[1] - local_center[1],
                      vertex[2] - local_center[2]};
    for (std::size_t axis = 0; axis != 3; ++axis)
      vertex[axis] = rotation[axis * 3] * local[0] +
                     rotation[axis * 3 + 1] * local[1] +
                     rotation[axis * 3 + 2] * local[2] + center[axis];
  }
  return value;
}

auto make_planes(const tf::points_buffer<double, 3> &solid_points, int cuts)
    -> std::vector<storage> {
  const auto [lower_bound, upper_bound] = examples::point_bounds(solid_points);
  const point middle{(lower_bound[0] + upper_bound[0]) / 2.0,
                     (lower_bound[1] + upper_bound[1]) / 2.0,
                     (lower_bound[2] + upper_bound[2]) / 2.0};
  std::vector<storage> planes;
  if (cuts > 0)
    planes.reserve(static_cast<std::size_t>(cuts * 3));
  for (int axis = 0; axis != 3; ++axis) {
    point normal{0.0, 0.0, 0.0};
    normal[static_cast<std::size_t>(axis)] = 1.0;
    for (int index = 0; index < cuts; ++index) {
      const auto t =
          static_cast<double>(index + 1) / static_cast<double>(cuts + 1);
      auto center = middle;
      center[static_cast<std::size_t>(axis)] =
          lower_bound[static_cast<std::size_t>(axis)] +
          t * (upper_bound[static_cast<std::size_t>(axis)] -
               lower_bound[static_cast<std::size_t>(axis)]);
      planes.push_back(make_plane(center, normal));
    }
  }
  return planes;
}

auto cell_color(std::size_t index) -> std::array<double, 3> {
  const auto &value = palette[(index * 5) % palette.size()];
  return {value[0] / 255.0, value[1] / 255.0, value[2] / 255.0};
}

auto point_string(const double *value) -> std::string {
  std::ostringstream stream;
  stream << std::fixed << std::setprecision(4) << '(' << value[0] << ", "
         << value[1] << ", " << value[2] << ')';
  return stream.str();
}

class camera_interactor : public vtkInteractorStyleTrackballCamera {
public:
  static auto New() -> camera_interactor *;
  vtkTypeMacro(camera_interactor, vtkInteractorStyleTrackballCamera);

  auto initialize(vtkCamera *camera, vtkRenderWindow *window) -> void {
    _camera = camera;
    _window = window;
  }

  auto OnKeyPress() -> void override {
    const auto key = std::string(this->Interactor->GetKeySym());
    if (key == "c" || key == "C") {
      const auto *size = _window->GetSize();
      std::cout << "\ncamera_settings{\n"
                << "  " << point_string(_camera->GetPosition()) << ",\n"
                << "  " << point_string(_camera->GetFocalPoint()) << ",\n"
                << "  " << point_string(_camera->GetViewUp()) << ",\n"
                << "  " << std::fixed << std::setprecision(3)
                << _camera->GetViewAngle() << ", {" << size[0] << ", "
                << size[1] << "}}\n\n";
      return;
    }
    vtkInteractorStyleTrackballCamera::OnKeyPress();
  }

private:
  vtkCamera *_camera = nullptr;
  vtkRenderWindow *_window = nullptr;
};

vtkStandardNewMacro(camera_interactor);

void set_camera(vtkRenderer *renderer, vtkRenderWindow *window,
                vtkCamera *camera) {
  if (camera_override) {
    camera->SetPosition(camera_override->position.data());
    camera->SetFocalPoint(camera_override->focal_point.data());
    camera->SetViewUp(camera_override->view_up.data());
    camera->SetViewAngle(camera_override->view_angle);
    window->SetSize(camera_override->size[0], camera_override->size[1]);
  } else {
    window->SetSize(1400, 1400);
    camera->SetViewUp(0.0, 0.0, 1.0);
    camera->SetFocalPoint(0.0, 0.0, 0.0);
    camera->SetPosition(0.55, -1.0, 0.32);
    renderer->ResetCamera();
    camera->Zoom(1.35);
  }
  renderer->ResetCameraClippingRange();
}

void add_cell_actors(vtkRenderer *renderer, const storage &cell,
                     std::size_t index) {
  auto poly = examples::to_vtk_polydata(cell);
  auto mapper = vtkSmartPointer<vtkPolyDataMapper>::New();
  mapper->SetInputData(poly);
  auto actor = vtkSmartPointer<vtkActor>::New();
  actor->SetMapper(mapper);
  auto *property = actor->GetProperty();
  const auto fill_color = cell_color(index);
  property->SetColor(fill_color[0], fill_color[1], fill_color[2]);
  property->SetEdgeVisibility(0);
  property->SetInterpolationToPhong();
  property->SetAmbient(0.32);
  property->SetDiffuse(0.68);
  property->SetSpecular(0.18);
  property->SetSpecularPower(28.0);
  renderer->AddActor(actor);

  auto feature_edges = vtkSmartPointer<vtkFeatureEdges>::New();
  feature_edges->SetInputData(poly);
  feature_edges->BoundaryEdgesOn();
  feature_edges->FeatureEdgesOn();
  feature_edges->SetFeatureAngle(25.0);
  feature_edges->ManifoldEdgesOff();
  feature_edges->NonManifoldEdgesOff();
  feature_edges->Update();
  auto edge_mapper = vtkSmartPointer<vtkPolyDataMapper>::New();
  edge_mapper->SetInputConnection(feature_edges->GetOutputPort());
  edge_mapper->ScalarVisibilityOff();
  auto edge_actor = vtkSmartPointer<vtkActor>::New();
  edge_actor->SetMapper(edge_mapper);
  edge_actor->GetProperty()->SetColor(edge_rgb[0], edge_rgb[1], edge_rgb[2]);
  edge_actor->GetProperty()->SetLineWidth(1.1);
  edge_actor->GetProperty()->LightingOff();
  renderer->AddActor(edge_actor);
}

} // namespace

int main(int argc, char **argv) {
  try {
    const auto parsed = parse_options(argc, argv);
    if (!parsed)
      return 0;
    const auto &arguments = *parsed;

    const auto solid = load_closed(arguments.mesh_path);
    const auto name = arguments.mesh_path.filename().string();
    const auto [lower_bound, upper_bound] =
        examples::point_bounds(solid.points_buffer());
    std::cout << name << ": " << solid.size() << " tris, bbox ["
              << lower_bound[0] << ' ' << lower_bound[1] << ' '
              << lower_bound[2] << "] .. [" << upper_bound[0] << ' '
              << upper_bound[1] << ' ' << upper_bound[2] << "]\n";

    const auto planes = make_planes(solid.points_buffer(), arguments.cuts);

    // The graph reads its operands for as long as it lives: the storage and
    // the caches below outlive it, and every cut is one more reading.
    std::vector<cache_t> caches(planes.size() + 1);
    std::vector<mesh_t> operands;
    operands.reserve(caches.size());
    operands.push_back(examples::mesh_over(solid, caches.front()));
    for (std::size_t plane = 0; plane != planes.size(); ++plane)
      operands.push_back(examples::mesh_over(planes[plane], caches[plane + 1]));

    const auto graph = tf::cpp::make_csg_graph(std::move(operands));
    constexpr auto domain_config = tf::domain_config::exclude_outer_shell |
                                   tf::domain_config::ignore_open_fragments;
    auto domains =
        arguments.all_domains
            ? tf::cpp::make_csg_domains(graph, domain_config)
            : tf::cpp::make_csg_domains(graph, tf::csg::op(0), domain_config);
    std::cout << "cells: " << domains.meshes.size()
              << (arguments.all_domains ? "  (ALL)" : "  (interior of solid)")
              << '\n';

    const auto scene_center = examples::point_mean(solid.points_buffer());
    std::vector<storage> cells;
    cells.reserve(domains.meshes.size());
    for (auto &cell : domains.meshes) {
      if (cell.points_buffer().size() == 0 || cell.size() < 1)
        continue;
      const auto center = examples::point_mean(cell.points_buffer());
      for (auto vertex : cell.points())
        for (std::size_t axis = 0; axis != 3; ++axis)
          vertex[axis] +=
              arguments.explode * (center[axis] - scene_center[axis]);
      cells.push_back(std::move(cell));
    }
    std::cout << "rendered domains: " << cells.size() << '\n';

    auto renderer = vtkSmartPointer<vtkRenderer>::New();
    renderer->SetBackground(1.0, 1.0, 1.0);
    if (!arguments.png)
      renderer->UseFXAAOn();
    for (std::size_t index = 0; index != cells.size(); ++index)
      add_cell_actors(renderer, cells[index], index);

    auto light_kit = vtkSmartPointer<vtkLightKit>::New();
    light_kit->SetKeyLightIntensity(0.9);
    light_kit->SetKeyToFillRatio(2.2);
    light_kit->SetKeyToHeadRatio(3.5);
    light_kit->SetKeyToBackRatio(3.0);
    light_kit->AddLightsToRenderer(renderer);

    auto window = vtkSmartPointer<vtkRenderWindow>::New();
    window->SetMultiSamples(8);
    window->AddRenderer(renderer);
    const auto title = std::string("csg_fracture: ") + name + " / " +
                       std::to_string(cells.size()) + " cells / " +
                       std::to_string(arguments.cuts) + " cuts";
    window->SetWindowName(title.c_str());
    auto *camera = renderer->GetActiveCamera();
    set_camera(renderer, window, camera);

    if (arguments.png) {
      window->SetAlphaBitPlanes(1);
      window->SetOffScreenRendering(1);
      window->Render();
      auto image = vtkSmartPointer<vtkWindowToImageFilter>::New();
      image->SetInput(window);
      image->SetInputBufferTypeToRGBA();
      image->ReadFrontBufferOff();
      image->Update();
      auto writer = vtkSmartPointer<vtkPNGWriter>::New();
#if defined(_WIN32)
      const auto png_path =
          std::filesystem::temp_directory_path() / "fracture.png";
#else
      const auto png_path = std::filesystem::path("/tmp/fracture.png");
#endif
      const auto png_path_string = png_path.string();
      writer->SetFileName(png_path_string.c_str());
      writer->SetInputConnection(image->GetOutputPort());
      writer->Write();
      if (writer->GetErrorCode() != vtkErrorCode::NoError)
        throw std::runtime_error(
            "failed to write " + png_path_string + ": " +
            vtkErrorCode::GetStringFromErrorCode(writer->GetErrorCode()));
      std::cout << "wrote " << png_path_string << " (RGBA)\n";
    } else {
      auto interactor = vtkSmartPointer<vtkRenderWindowInteractor>::New();
      interactor->SetRenderWindow(window);
      auto style = vtkSmartPointer<camera_interactor>::New();
      style->initialize(camera, window);
      interactor->SetInteractorStyle(style);
      window->Render();
      std::cout << "interactive: drag orbit, scroll zoom, 'c' print camera, "
                   "'q' quit\n";
      interactor->Start();
    }
    return 0;
  } catch (const std::exception &error) {
    std::cerr << "error: " << error.what() << '\n';
    print_usage(argv[0], std::cerr);
    return 1;
  }
}
